Bluetooth: gatt settings (NVS) fixes

jori-nordic opened the pull request:
See commit messages.

I will also write a bsim test for this, hence why the PR is in draft.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/54172.
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/54173.

jori-nordic committed:
Bluetooth: host: Store Client Supported Features on write

Store the Client supported features value when it is written to, instead of
only on disconnection/identity resolved.

Works around the situation where a user bonds, CF is written and the device
is abruptly powered off, discarding the CF value, but keeping the bond.

Fixes #54172.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: host: store GATT change-aware status in settings

The spec says we have to persist the change-aware status of bonded peers
between resets.
This stores it at the end of the CF storage that we currently have.

Fixes #54173

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1256,115 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
 {
-	if ((!gatt_ccc_conn_is_queued(conn)) &&
-	    bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+	if (!conn) {
```
there might be a bug here:
- if the device is disconnected, and we (de)register a gatt service afterwards, we won't store the change-aware status
- also need to double-check there is no way `_dequeue` will get called with `conn == NULL`.

jori-nordic replied:
it looks like the first point isn't really an issue: if we don't get powered down, then the change-aware status will be correct (it's just not stored to NVS).
If we power-down, then the gatt db will be reset to its initial state, and if we decide to alter the db, the peer will be marked correctly as change-unaware when the connection is established (and will refuse ATT operations)

jori-nordic replied:
alright, on second thought, this is a bad design. I wil instead try to rework `bt_gatt_store_cf` to not need `conn` (seems easy) so we always store to NVS regardless of connection status.

jori-nordic replied:
@MarekPieta you can take another look now

MarekPieta replied:
Yes, I just started rereview

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1256,115 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
+	bool bonded = bt_addr_le_is_bonded(conn->id, &conn->le.dst);
+
+	if (!queued && bonded) {
 		/* Store the connection with the same index it has in
```
I don't really like this, leads to problems when doing stuff to the gatt db when the connection is not active.

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1225,109 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
 
-		k_work_reschedule(&gatt_ccc_store.work, CCC_STORE_DELAY);
+		if (el->conn == NULL) {
+			el->conn = bt_conn_ref(conn);
```
Btw. should we somehow ensure that all flags are cleared just after we allocated the new connection? Otherwise there could be a risk that new conn inherits flags from previous one.

jori-nordic committed:
Bluetooth: host: don't overwrite GATT DB hash before settings_load

Previously, if the app registered a bunch of services at boot before
calling `settings_load()`, then the hash would be silently overwritten.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic committed:
Bluetooth: host: add test for some GATT settings bugs

This test is designed to check the soundness of GATT caching across
reboots.

It will fail if the fixes for #54173 and #54172 are not present.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5896,13 +5926,30 @@ static int cf_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+		if (len != sizeof(dst)) {
+			LOG_DBG("Change-aware status not stored, "
+				"defaulting peer status to change-aware");
+			set_change_aware(NULL, cfg, true);
```
Good point, I reverted it to change-unaware.

For the bit, I prefer if that bit is set from only one place. 
I removed the manual setting below, since in that case (when running `settings_load`) we will not be connected, and `set_change_aware` will only set the bit and not try to write it back to NVS.

`set_change_aware` needs to be able to have `conn == NULL` because it can get called from other places where the connection is not valid anymore.

jori-nordic commented:
Moving out of draft. The stack changes / test can be reviewed, I'll fix the CI errors on the next push.

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1225,110 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
-		k_work_reschedule(&gatt_ccc_store.work, CCC_STORE_DELAY);
+		if (el->conn == NULL) {
+			el->conn = bt_conn_ref(conn);
+			atomic_set(el->flags, 0);
```
There is hidden a type error here. `atomic_set` takes a pointer to a single `atomic_t`. `el->flags` is an array of `atomic_t`. This happens to work fine so long as `enum delayed_store_flags` has fewer than `ATOMIC_BITS` members, which admittedly is quite large; so maybe this case is fine by our code standard since we don't anticipate the flag count to grow that much?

jori-nordic replied:
yeah I don't think we're gonna cross `ATOMIC_BITS`, is that 32 or something?

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1225,110 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
-			bt_conn_unref(conn);
+			if (IS_ENABLED(CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE) &&
+			    gatt_delayed_store_is_queued(conn, DELAYED_STORE_CCC)) {
+				bt_gatt_store_ccc(conn->id, &conn->le.dst);
+				gatt_delayed_store_dequeue(conn, DELAYED_STORE_CCC);
```
I think these effects must be done in reverse order. The flag must be cleared first. When the flag is set again while the store-procedure is in progress (possibly almost finished), it is necessary to run the procedure again.

alwa-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -208,15 +208,40 @@ config BT_SETTINGS_CCC_LAZY_LOADING
+	# Enables delayed non-volatile storage mechanism
+	bool
+	help
+	  Triggers the storage of the CF and CCC right after a write.
```
What happens when you set this config to false?

jori-nordic replied:
this is a convenience option, not user-selectable. Check out the `CCC_STORE_ON_WRITE` and `CF_STORE_ON_WRITE` help strings for what happens in that case.

alwa-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -208,15 +208,40 @@ config BT_SETTINGS_CCC_LAZY_LOADING
+	# Enables delayed non-volatile storage mechanism
+	bool
+	help
+	  Triggers the storage of the CF and CCC right after a write.
```
What happens when you set this config to false?

alwa-nordic replied:
What happens when you set this config to false?

alwa-nordic commented on tests/bluetooth/bsim_bt/bsim_test_gatt_settings/src/server.c:
```diff
@@ -0,0 +1,77 @@
+		set_security(conn, BT_SECURITY_L2);
+	}
+
+	wait_disconnected();
```
This doesn't test that the write happens before the disconnection.

alwa-nordic replied:
But also: Does anything bad happen if we power-off in the on-disconnected callback? (I suspect some work items will not run.)

jori-nordic replied:
I will change that so the server dies before the connection is ended.
For the second point, I think that should rather be in a separate bsim robustness-type test, since I suppose other stuff than GATT might break.

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -549,10 +553,58 @@ enum delayed_store_flags {
+
+#if defined(CONFIG_BT_SETTINGS_DELAYED_STORE)
+	if (changed) {
+		gatt_delayed_store_enqueue(cfg->id, &cfg->peer, DELAYED_STORE_CF);
```
With the current approach we may need to increase number of delayed CF writes that can be queued (to `(CONFIG_BT_MAX_PAIRED + CONFIG_BT_MAX_CONN)`). Keep in mind that we may call `set_change_aware` also for bonded peers that are not connected. Alternatively maybe we could simply store the value instantly if peer is not connected?

jori-nordic replied:
I'd prefer the second solution, but it seems like an optimization. I'd rather do it after getting the bug fix in.

MarekPieta replied:
The count seems updated, so it should work fine :)

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1228,135 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
+				bt_gatt_store_cf(el->id, &el->peer);
+			}
+
+			if (atomic_get(el->flags) == 0) {
+				gatt_delayed_store_free(el);
+			}
```
Not sure if we need to do it with current implementation. During allocation we can override the `gatt_delayed_store` object if flags are set to zeros anyway

jori-nordic replied:
let's do it just for correctness.

MarekPieta commented:
During the review I focused on implementation in `gatt.c` file. I am not too familiar with `bsim`.

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -878,8 +878,11 @@ static void sc_indicate(uint16_t start, uint16_t end);
+
 #if defined(CONFIG_BT_SETTINGS)
-	if (atomic_test_and_clear_bit(gatt_sc.flags, DB_HASH_LOAD)) {
+	store = atomic_test_and_clear_bit(gatt_sc.flags, DB_HASH_LOAD);
```
yes, in the case `store` = false. The prev. behavior was to always store, even if we hadn't actually loaded the hash from the NVS.

MarekPieta replied:
Are you sure? It seems that the value of `store` is not used anywhere undef if and code under if ends with return. If BT_SETTINGS are disabled, the value of store is set to `true` anyway so behavior seems not changed.

MarekPieta replied:
There is an additional `db_hash_store` under the `if` statement too

jori-nordic replied:
this is inside the ifdef: 
![image](https://user-images.githubusercontent.com/68428270/218031755-e87bf970-9a08-4589-9d21-da911bdbd9f2.png)

jori-nordic replied:
if the settings haven't yet been loaded, `store` will be false. previously we would then call `db_hash_gen` (at the end of this fn) with store=true, overwriting the value in the settings. 
Then when settings are loaded, the hash is recalculated, but it will match with the one in NVS (since it has just been overwritten).

MarekPieta replied:
Ahh, you are right. There is a case where we do not enter `if` and it makes the difference. I initially overlooked it - sorry.

jori-nordic replied:
so there was actually a bug here, I reworked it.

MarekPieta replied:
Yes, thanks :)

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1228,137 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
-			bt_conn_ref(conn);
+	if (el) {
+		memset(&el->peer, 0, sizeof(el->peer));
+		atomic_set(el->flags, 0);
```
since 0 is a valid `id`, I didn't bother. But yeah I can do that.

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1228,137 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
+	if (bonded) {
+		if (el == NULL) {
+			el = gatt_delayed_store_alloc(id, peer_addr);
+			if (el == NULL) {
```
Yeah I also think it shouldn't happen.

jori-nordic commented on tests/bluetooth/bsim_bt/bsim_test_gatt_settings/src/client.c:
```diff
@@ -0,0 +1,105 @@
+	printk("start round 0...........\n");
+
+	conn = connect_as_peripheral();
+	printk("connected: conn %p\n", conn);
```
this is a common idiom in zephyr (that we don't explicitly cast to void), take a look at e.g. `gatt.c`

MarekPieta replied:
C standard states that `for %p the argument shall be a pointer to void.`
Anyway let's leave it if it's acceptable here.

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5927,13 +5984,34 @@ static int cf_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+		if (len != sizeof(dst)) {
+			LOG_WRN("Change-aware status not found in settings, "
+				"defaulting peer status to change-unaware");
+			set_change_aware(cfg, false);
```
I added a check for change-awareness once the gatt bearer has been encrypted (`bt_gatt_encrypt_change`), seems to fit the bill. I added a test round to double-check that we do receive the SC indication when the change-unaware status is loaded from the settings.

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5927,13 +5984,34 @@ static int cf_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+				LOG_WRN("Read back bad change-aware value: 0x%x, "
+					"defaulting peer status to change-unaware",
+					change_aware);
+				set_change_aware(cfg, false);
```
should be handled by this https://github.com/zephyrproject-rtos/zephyr/pull/54349/files#r1101366188

hermabe commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1211,73 +1300,129 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
+
+		atomic_set_bit(el->flags, flag);
+
+		k_work_reschedule(&gatt_delayed_store.work,
+				  K_MSEC(CONFIG_BT_SETTINGS_DELAYED_STORE_MS));
```
Nothing will be stored to flash if the device reboots/loses power before `CONFIG_BT_SETTINGS_DELAYED_STORE_MS`?

MarekPieta replied:
That's true. Proper fix would require to write from RX thread in some places I think (otherwise we never know if the value would actually be stored before response is sent). I noticed that implementation tries to avoid it.

hermabe replied:
Yeah, not sure if there is a better way to solve it. It can anyway wait until laters as it was not introduced here as you say.

hermabe commented on tests/bluetooth/bsim_bt/bsim_test_gatt_settings/src/gatt_utils.c:
```diff
@@ -0,0 +1,368 @@
+/* We need to discover:
+ * - Dynamic service
+ * - Client Features (to set robust caching)
+ * - DB hash handle (to get out of the change-unaware state)
```
I dont see the DB hash is read anywhere? Also a thing to note, the DB hash shall use read by uuid:
> In order to read the value of this characteristic the client shall always use the
GATT Read Using Characteristic UUID sub-procedure. The Starting Handle
should be set to 0x0001 and the Ending Handle should be set to 0xFFFF.

Not sure it is worth caring about in a test.

jori-nordic replied:
I was reading it in a previous version. I'll remove it.

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -852,6 +904,8 @@ static void db_hash_process(struct k_work *work)
 		 */
 		sc_indicate(0x0001, 0xffff);
 
+		set_all_change_unaware();
```
There is an edge case I think. It could appear right after the DFU.

We see that database hash does not match, so we store new hash and set change awareness to false (keep in mind the change awareness store operation is delayed). At this point we may be rebooted and after the reboot we no longer see that the database hash was modified (so we forgotten that peers should have been marked as change unaware).

Should not have high impact anyway since we would most probably send SC indications.

MarekPieta replied:
There is an edge case I think. It could appear right after the DFU.

We see that database hash does not match, so we store new hash and set change awareness to false (keep in mind the change awareness store operation is delayed). At this point we may be rebooted and after the reboot we no longer see that the database hash was modified (so we forgotten that peers should have been marked as change unaware).

Should not have high impact anyway since we would most probably send SC indications.

jori-nordic replied:
hm that is true, I did misunderstand you. 
The best way would be to store a "last seen hash" for each peer. I would do that if storage was free. 
I've changed it now to store to settings in `set_all_change_unaware`.

MarekPieta replied:
Should be good enough.
I think it's better to force a few unnecessary discoveries than to miss a necessary discovery.

jori-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -502,6 +500,9 @@ static struct _bt_gatt_ccc sc_ccc = BT_GATT_CCC_INITIALIZER(NULL,
 							    sc_ccc_cfg_write,
 							    NULL);
 
+/* Do not shuffle the values in this enum, they are used as bit offsets when
+ * saving the CF flags to NVS (i.e. NVS persists between FW upgrades).
+ */
```
I don't think there is. I'm kinda already doing this when reading it (e.g. if the bit is not present or a bad value).

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5547,6 +5717,15 @@ void bt_gatt_encrypt_change(struct bt_conn *conn)
+		/* Send a Service Changed indication if the current peer is
+		 * marked as change-unaware.
+		 */
+		sc_indicate(0x0001, 0xffff);
```
We could consider optimizing it and indicating only to this particular peer (may be done in the future since it should not lead to bugs - it's just an optimization)

jori-nordic replied:
yeah I wanted to do that originally, but its trickier than it seems because we use global flags, not per-peer for book-keeping. Maybe will do the optimization later.

jori-nordic replied:
yep, will do that in the future. Please open an Enhancement issue linking this comment, you can assign it to me :+1:

MarekPieta replied:
https://github.com/zephyrproject-rtos/zephyr/issues/54766

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -638,7 +704,13 @@ static ssize_t cf_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
 	bt_addr_le_copy(&cfg->peer, &conn->le.dst);
 	cfg->id = conn->id;
-	atomic_set_bit(cfg->flags, CF_CHANGE_AWARE);
+	set_change_aware(cfg, true);
```
If a connected peer writes CF value before bonded and then reboots without disconnection, the written CF value is lost. On the other hand if the same peer disconnects before the reboot, the CF value is written to settings. Seems like bug.
Replicated this one with a Windows 10 used as Bluetooth Central.

MarekPieta replied:
btw. Similar issue could affect CCC values if they are written before bonded.
Maybe we should store CCC and CF also on `pairing_complete` from `bt_conn_auth_info_cb `  to prevent the issue (I think that could be done in a separate PR)?

jori-nordic replied:
Yeah I think we should only ever store to NVS when we are sure that we are bonded. I will double-check the spec to be sure.

MarekPieta replied:
Discussed offline. We do not store to NVS when we are not bonded (we already have proper checks in code to ensure this).

The main problem is that we may not store the CCC and CF values if peer writes it and then bonds. We need to trigger write when bonding is complete (the `identity_resolved` callback would not be called if the connected peer does not use privacy).

jori-nordic replied:
can you please open another bug and assign me?

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -885,7 +886,28 @@ static void sc_indicate(uint16_t start, uint16_t end);
+		/* we want to generate the hash, but not overwrite the hash
+		 * stored in settings, that we haven't yet loaded.
+		 */
+		db_hash_gen(false);
```
can we get `db_hash_store` out of `db_hash_gen()`?

jori-nordic replied:
done

alwa-nordic reviewed (changes requested):
&ZeroWidthSpace;

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -638,7 +704,13 @@ static ssize_t cf_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
-	atomic_set_bit(cfg->flags, CF_CHANGE_AWARE);
+	set_change_aware(cfg, true);
+
+#if defined(CONFIG_BT_SETTINGS_DELAYED_STORE)
+	/* Save the new configuration to NVM (in another thread) */
+	LOG_DBG("trigger CF write");
+	gatt_delayed_store_enqueue(conn->id, &conn->le.dst, DELAYED_STORE_CF);
+#endif
```
Isn't it already done by `set_change_aware` now?

MarekPieta commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -813,7 +886,28 @@ static void sc_indicate(uint16_t start, uint16_t end);
 static void db_hash_process(struct k_work *work)
 {
 #if defined(CONFIG_BT_SETTINGS)
```
Maybe we could perform `db_hash_process` in `db_hash_commit` callback context (instead of moving it to Bluetooth long workqueue context that has low priority)?
That could allow us to avoid synchronization issues related to processing the db_hash and marking peers as change unaware when GATT database changes after a DFU operation. I think current implementation introduces possibility of accessing GATT by previously bonded peer before DB hash is processed and before the peer is properly marked as change unaware.
The issue also seem not directly related to the PR so may be addressed separately I think.

jori-nordic replied:
Yeah, please open another low-prio bug for that. you can assign me for now as well.
I think it should be okay to NAK all GATT access until the hash has been calculated.

MarekPieta replied:
Alternatively, maybe we could delay sending GATT response and force the database hash calculation in the current context (I am not sure which solution would provide better interoperability with devices available on the market)

jori-nordic replied:
that may be a valid idea, but this is kinda why we added this long-wq in the first place: bt runs as a coop prio, and the hash can be quite long, and will pre-empt other threads from running (even if cooperative).

MarekPieta replied:
Follow-up issue: https://github.com/zephyrproject-rtos/zephyr/issues/54773

MarekPieta reviewed (dismissed):
One small improvement: https://github.com/zephyrproject-rtos/zephyr/pull/54349#discussion_r1102865765

Also added 2 comments related to other possible issues found during the review. Maybe we could address these issues in separate PRs to limit scope of this one?

jori-nordic committed:
Bluetooth: host: don't store hash in db_hash_gen()

Move out the storage of the calculated hash from the fn that calculates
it (db_hash_gen).

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jhedberg commented:
Seems like this might be good to get included in the 3.3 release.

stephanosio reviewed (commented):
The changes in this PR are confined to the Bluetooth subsystem and, given that there is a dedicated team vigorously testing and fixing any Bluetooth-related bugs, I am ok with including this in the 3.3 release.
